### PR TITLE
GKE integration

### PR DESCRIPTION
**Pull request by colin014**
- create cluster
- delete cluster
- get cluster
- cluster status
- fetch config
- posthook functions

### CODE REVIEW

File: cloud/base.go [hunk 1]
```diff
@@ -254,6 +270,22 @@ func GetK8SConfig(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context) ([]byte,
 			config, _ := base64.StdEncoding.DecodeString(b64config.Properties.KubeConfig)
 			return config, nil
 		}
+	case banzaiConstants.Google:
+		banzaiUtils.LogInfo(LOGTAG, "Trying to get GoogleKubernetesConfig")
+		config, err := getGoogleKubernetesConfig(cs)
```
Comment (baluchicken) [inline cloud/base.go]: I think we should return byte[] here.

File: cloud/base.go [hunk 2]
```diff
@@ -254,6 +270,22 @@ func GetK8SConfig(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context) ([]byte,
...
+			// something went wrong
+			banzaiUtils.LogWarn(LOGTAG, "Error during getting Google K8S config")
+			SetResponseBodyJson(c, err.StatusCode, gin.H{
+				JsonKeyStatus: err.StatusCode,
+				JsonKeyData:   err.Message,
+			})
+			return nil, errors.New("error happened during getting K8S config")
+		} else {
+			banzaiUtils.LogInfo(LOGTAG, "Kubernetes Config retrieve succeeded!")
+			c := to.String(config)
```
Comment (baluchicken) [inline cloud/base.go]: So we don't need to convert this to byte[]

File: cloud/gke.go [hunk 3]
```diff
@@ -0,0 +1,1038 @@
...
+			default:
+				banzaiUtils.LogDebug(banzaiConstants.TagFetchClusterConfig, "Content-Type: ", ctype)
+				SetResponseBodyString(c, http.StatusOK, encodedConfig)
+			}
+		}
+	}
+
+}
+
+func getGoogleKubernetesConfig(cs *banzaiSimpleTypes.ClusterSimple) (*string, *banzaiTypes.BanzaiResponse) {
```
Comment (baluchicken) [inline cloud/gke.go]: I think we should return a byte[] here.

File: cloud/gke.go [hunk 4]
```diff
@@ -0,0 +1,1038 @@
...
+const (
+	defaultNamespace = "default"
+	clusterAdmin     = "cluster-admin"
+	netesDefault     = "netes-default"
+)
+
+const googleAppCredential = "GOOGLE_APPLICATION_CREDENTIALS"
+
+func init() {
+	credentialPath = os.Getenv(googleAppCredential)
```
Comment (baluchicken) [inline cloud/gke.go]: Shouldn't we use viper.getString here? and put GOOGLE_APPLICATION_CREDENTIALS to config.toml.example?

File: cloud/gke.go [hunk 5]
```diff
@@ -0,0 +1,1038 @@
...
+	banzaiUtils.LogDebugf(banzaiConstants.TagInit, "%s is %s", googleAppCredential, credentialPath)
+}
+
+func CreateClusterGoogle(request *banzaiTypes.CreateClusterRequest, c *gin.Context) (bool, *banzaiSimpleTypes.ClusterSimple) {
+	data, err := ioutil.ReadFile(credentialPath)
+	if err != nil {
+		banzaiUtils.LogFatalf(banzaiConstants.TagCreateCluster, "GOOGLE_APPLICATION_CREDENTIALS env var is not specified: %s", err)
```
Comment (baluchicken) [inline cloud/gke.go]: Here this can be an error. Please use %v with err or %s with err.Error()

File: cloud/gke.go [hunk 6]
```diff
@@ -0,0 +1,1038 @@
...
+		Zone:              request.Location,
+		Name:              request.Name,
+		NodeCount:         int64(request.Properties.CreateClusterGoogle.Node.Count),
+		CredentialPath:    credentialPath,
+		CredentialContent: string(data),
+		MasterVersion:     request.Properties.CreateClusterGoogle.Master.Version,
+		NodeVersion:       request.Properties.CreateClusterGoogle.Node.Version,
+	}
+
+	banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Cluster request: %v", generateClusterCreateRequest(cc))
```
Comment (baluchicken) [inline cloud/gke.go]: Maybe this is nothing but why are we calling generateClusterCreateRequest(cc) three times here?

File: cloud/gke.go [hunk 7]
```diff
@@ -0,0 +1,1038 @@
...
+	createCall, err := svc.Projects.Zones.Clusters.Create(cc.ProjectID, cc.Zone, generateClusterCreateRequest(cc)).Context(context.Background()).Do()
+
+	banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Cluster request submitted: %v", generateClusterCreateRequest(cc))
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
+		banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Contains error: %s", err)
```
Comment (baluchicken) [inline cloud/gke.go]: I assume this one is a LogErrorf and again I think %s requires err.Error() and %v with err

File: cloud/gke.go [hunk 8]
```diff
@@ -0,0 +1,1038 @@
...
+			JsonKeyMessage: be.Message,
+		})
+		return false, nil
+	} else {
+		banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Cluster %s create is called for project %s and zone %s. Status Code %v", cc.Name, cc.ProjectID, cc.Zone, createCall.HTTPStatusCode)
+	}
+
+	gkeCluster, err := waitForCluster(svc, cc)
+	if err != nil {
+		banzaiUtils.LogErrorf(banzaiConstants.TagCreateCluster, "Cluster create failed", err)
```
Comment (baluchicken) [inline cloud/gke.go]: This one is wrong %v missing

File: cloud/gke.go [hunk 9]
```diff
@@ -0,0 +1,1038 @@
...
+		Google: banzaiSimpleTypes.GoogleClusterSimple{
+			Project:       request.Properties.CreateClusterGoogle.Project,
+			NodeCount:     request.Properties.CreateClusterGoogle.Node.Count,
+			MasterVersion: request.Properties.CreateClusterGoogle.Master.Version,
+			NodeVersion:   request.Properties.CreateClusterGoogle.Node.Version,
+		},
+	}
+
+	if err := database.Save(&cluster2Db).Error; err != nil {
+		DbSaveFailed(c, err, cluster2Db.Name)
```
Comment (baluchicken) [inline cloud/gke.go]: Maybe it is nothing, but here we are swallowing an error

File: cloud/gke.go [hunk 10]
```diff
@@ -0,0 +1,1038 @@
...
+		JsonKeyStatus:     http.StatusCreated,
+		JsonKeyResourceId: cluster2Db.ID,
+		JsonKeyData:       gkeCluster,
+	})
+	return true, &cluster2Db
+}
+
+func getBanzaiErrorFromError(err error) *banzaiTypes.BanzaiResponse {
+
+	if err == nil {
```
Comment (baluchicken) [inline cloud/gke.go]: This code is unreachable, because every time this method is called it is inside a if err != nil

File: cloud/gke.go [hunk 11]
```diff
@@ -0,0 +1,1038 @@
...
+	if err != nil {
+		banzaiUtils.LogFatalf(banzaiConstants.TagCreateCluster, "Could not initialize gke client: %v", err)
+		return nil, err
+	}
+	banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Using service acc: %v", service)
+	return service, nil
+}
+
+// Struct of GKE
+type GKECluster struct {
```
Comment (baluchicken) [inline cloud/gke.go]: Maybe this can be put inside banzai-types.

File: cloud/gke.go [hunk 12]
```diff
@@ -0,0 +1,1038 @@
...
+	LegacyAbac bool
+	// NodePool id
+	NodePoolID string
+	// Image Type
+	ImageType string
+}
+
+func waitForCluster(svc *gke.Service, cc GKECluster) (*gke.Cluster, error) {
+
+	message := ""
```
Comment (baluchicken) [inline cloud/gke.go]: maybe this can be replaced with var message string

File: cloud/gke.go [hunk 13]
```diff
@@ -0,0 +1,1038 @@
...
+	for {
+
+		cluster, err := svc.Projects.Zones.Clusters.Get(cc.ProjectID, cc.Zone, cc.Name).Context(context.TODO()).Do()
+		if err != nil {
+			return nil, err
+		}
+
+		if cluster.Status == statusRunning {
+			banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Cluster %v is running", cc.Name)
```
Comment (baluchicken) [inline cloud/gke.go]: %s

File: cloud/gke.go [hunk 14]
```diff
@@ -0,0 +1,1038 @@
...
+
+func GetClusterGoogle(svc *gke.Service, cc GKECluster) (*gke.Cluster, error) {
+	return svc.Projects.Zones.Clusters.Get(cc.ProjectID, cc.Zone, cc.Name).Context(context.TODO()).Do()
+}
+
+func ReadClusterGoogle(cs *banzaiSimpleTypes.ClusterSimple, svc *gke.Service) *ClusterRepresentation {
+	banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "Read google cluster with", cs.Name, "id")
+
+	if cs == nil {
+		banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "<nil> cluster")
```
Comment (baluchicken) [inline cloud/gke.go]: Is this really a info?

File: cloud/gke.go [hunk 15]
```diff
@@ -0,0 +1,1038 @@
...
+
+	gkec := GKECluster{
+		ProjectID: cs.Google.Project,
+		Name:      cs.Name,
+		Zone:      cs.Location,
+	}
+
+	response, err := GetClusterGoogle(svc, gkec)
+	if err != nil {
+		banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "Something went wrong under read:", err)
```
Comment (baluchicken) [inline cloud/gke.go]: Again is this really an info log?

File: cloud/gke.go [hunk 16]
```diff
@@ -0,0 +1,1038 @@
...
+
+type GoogleRepresentation struct {
+	GoogleCluster *gke.Cluster `json:"value,omitempty"`
+}
+
+func GetClusterInfoGoogle(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context) {
+	banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "Fetch aks cluster with name:", cs.Name, "in", cs.Azure.ResourceGroup, "resource group.")
+
+	if cs == nil {
+		banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "<nil> cluster")
```
Comment (baluchicken) [inline cloud/gke.go]: This one is definitely not an info log level:)

File: cloud/gke.go [hunk 17]
```diff
@@ -0,0 +1,1038 @@
...
+		ProjectID: cs.Google.Project,
+		Name:      cs.Name,
+		Zone:      cs.Location,
+	}
+
+	response, err := GetClusterGoogle(svc, gkec)
+	if err != nil {
+		// fetch failed
+		googleApiErr := getBanzaiErrorFromError(err)
+		banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "Status code:", googleApiErr.StatusCode)
```
Comment (baluchicken) [inline cloud/gke.go]: Again are you sure this is logInfo?

File: cloud/gke.go [hunk 18]
```diff
@@ -0,0 +1,1038 @@
...
+
+	gkec := GKECluster{
+		ProjectID: cs.Google.Project,
+		Name:      cs.Name,
+		Zone:      cs.Location,
+	}
+
+	response, err := GetClusterGoogle(svc, gkec)
+	if err != nil {
+		// fetch failed
```
Comment (baluchicken) [inline cloud/gke.go]: Log this, that fetch failed

File: cloud/gke.go [hunk 19]
```diff
@@ -0,0 +1,1038 @@
...
+		googleApiErr := getBanzaiErrorFromError(err)
+		banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "Status code:", googleApiErr.StatusCode)
+		banzaiUtils.LogInfo(banzaiConstants.TagGetCluster, "Error during get cluster details:", googleApiErr.Message)
+		SetResponseBodyJson(c, googleApiErr.StatusCode, googleApiErr)
+	} else {
+		// fetch success
```
Comment (baluchicken) [inline cloud/gke.go]: log this "fetch success"

File: cloud/gke.go [hunk 20]
```diff
@@ -0,0 +1,1038 @@
...
+
+}
+
+// UpdateClusterGoogleInCloud updates google cluster in cloud
+func UpdateClusterGoogleInCloud(r *banzaiTypes.UpdateClusterRequest, c *gin.Context, preCluster banzaiSimpleTypes.ClusterSimple) bool {
+
+	banzaiUtils.LogInfo(banzaiConstants.TagUpdateCluster, "Start updating cluster (google)")
+
+	if r == nil {
+		banzaiUtils.LogInfo(banzaiConstants.TagUpdateCluster, "<nil> update cluster")
```
Comment (baluchicken) [inline cloud/gke.go]: Loginfo again.

File: cloud/gke.go [hunk 21]
```diff
@@ -0,0 +1,1038 @@
...
+			Project:       preCluster.Google.Project,
+			NodeCount:     r.GoogleNode.Count,
+			MasterVersion: r.GoogleMaster.Version,
+			NodeVersion:   r.GoogleNode.Version,
+		},
+	}
+
+	svc, err := GetGoogleServiceClient()
+	if err != nil {
+		SetResponseBodyJson(c, http.StatusInternalServerError, gin.H{
```
Comment (baluchicken) [inline cloud/gke.go]: Log the error.

File: cloud/gke.go [hunk 22]
```diff
@@ -0,0 +1,1038 @@
...
+	cc := GKECluster{
+		MasterVersion: r.GoogleMaster.Version,
+		NodeVersion:   r.GoogleNode.Version,
+		NodeCount:     int64(r.GoogleNode.Count),
+	}
+
+	res, err := callUpdateClusterGoogle(svc, cc)
+	if err != nil {
+		googleApiErr := getBanzaiErrorFromError(err)
+		banzaiUtils.LogInfo(banzaiConstants.TagUpdateCluster, "Cluster update failed!", googleApiErr)
```
Comment (baluchicken) [inline cloud/gke.go]: Are you sure this is log info

File: cloud/gke.go (hunk 9 above)
Comment (colin014) [reply cloud/gke.go]: The DbSaveFailed function will logging and setting the response body with the error:
https://github.com/banzaicloud/pipeline/blob/master/cloud/base.go#L74

File: cloud/gke.go (hunk 10 above)
Comment (colin014) [reply cloud/gke.go]: It's a protection for later use

File: cloud/gke.go (hunk 11 above)
Comment (colin014) [reply cloud/gke.go]: GKECluster is only used inside pipeline. The requests use other structs.

File: cloud/gke.go (hunk 18 above)
Comment (colin014) [reply cloud/gke.go]: It's logged, but only info:
https://github.com/banzaicloud/pipeline/blob/google-integration/cloud/gke.go#L358
I'll fix it.

File: cloud/gke.go (hunk 21 above)
Comment (colin014) [reply cloud/gke.go]: If there was an error, the GetGoogleServiceClient method will log

File: cloud/gke.go [hunk 23]
```diff
@@ -653,11 +654,13 @@ func GetGoogleK8SConfig(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context) {
 
 func getGoogleKubernetesConfig(cs *banzaiSimpleTypes.ClusterSimple) ([]byte, *banzaiTypes.BanzaiResponse) {
 
+	banzaiUtils.LogInfo(banzaiConstants.TagFetchClusterConfig, "Get Service Client")
```
Comment (stoader) [inline cloud/gke.go]: The message should read "Get Google Service Client"

File: cloud/gke.go [hunk 24]
```diff
@@ -653,11 +654,13 @@ func GetGoogleK8SConfig(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context) {
...
 	svc, err := GetGoogleServiceClient()
 	if err != nil {
 		banzaiUtils.LogErrorf(banzaiConstants.TagFetchClusterConfig, "Error during get service client %v", err)
 		return nil, getBanzaiErrorFromError(err)
 	}
+	banzaiUtils.LogInfo(banzaiConstants.TagFetchClusterConfig, "Get Service Client success")
```
Comment (stoader) [inline cloud/gke.go]: The message should read "Get Google Service Client success"

File: cloud/gke.go [hunk 25]
```diff
@@ -1016,18 +1019,19 @@ func GetGoogleClusterStatus(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context)
...
 		return
 	}
 
-	banzaiUtils.LogInfo(banzaiConstants.TagGetClusterStatus, "Load google props from database")
+	banzaiUtils.LogInfo(banzaiConstants.TagGetClusterStatus, "Load Google props from database")
 
 	// load google props from db
 	database.SelectFirstWhere(&cs.Google, banzaiSimpleTypes.GoogleClusterSimple{ClusterSimpleId: cs.ID})
 
+	banzaiUtils.LogInfo(banzaiConstants.TagGetClusterStatus, "Get Service Client")
```
Comment (stoader) [inline cloud/gke.go]: The message should read "Get Google Service Client"

File: cloud/gke.go [hunk 26]
```diff
@@ -1037,6 +1041,7 @@ func GetGoogleClusterStatus(cs *banzaiSimpleTypes.ClusterSimple, c *gin.Context)
 		})
 		return
 	}
+	banzaiUtils.LogInfo(banzaiConstants.TagGetClusterStatus, "Get Service Client success")
```
Comment (stoader) [inline cloud/gke.go]: The message should read "Get Google Service Client success"

**Comment by matyix** [comment]
Added log change requests by @stoader

File: cloud/gke.go [hunk 27]
```diff
@@ -0,0 +1,1080 @@
...
+			banzaiUtils.LogError(banzaiConstants.TagUpdateCluster, "Contains error", err)
+			return nil, err
+		}
+	}
+	return updatedCluster, nil
+}
+
+func waitForNodePool(svc *gke.Service, cc GKECluster) error {
+	const TAG = "waitForNodePool"
+	message := ""
```
Comment (baluchicken) [inline cloud/gke.go]: Just a minor thing I think we should use var message string here instead of adding empty string.

File: cloud/gke.go [hunk 28]
```diff
@@ -0,0 +1,1080 @@
...
+		SetResponseBodyJson(c, http.StatusInternalServerError, gin.H{
+			JsonKeyStatus:  http.StatusInternalServerError,
+			JsonKeyMessage: err,
+		})
+		return false
+	}
+	banzaiUtils.LogInfo(banzaiConstants.TagDeleteCluster, "Get Google Service Client succeeded")
+
+	banzaiUtils.LogInfof(banzaiConstants.TagDeleteCluster, "Removing cluster %v from project %v, zone %v", cc.Name, cc.ProjectID, cc.Zone)
+	deleteCall, err := svc.Projects.Zones.Clusters.Delete(cc.ProjectID, cc.Zone, cc.Name).Context(context.Background()).Do()
```
Comment (baluchicken) [inline cloud/gke.go]: I think this can be replaced with function GetClusterGoogle()

File: cloud/gke.go [hunk 29]
```diff
@@ -0,0 +1,1080 @@
...
+	banzaiUtils.LogInfo(banzaiConstants.TagFetchClusterConfig, "Get Google Service Client")
+	svc, err := GetGoogleServiceClient()
+	if err != nil {
+		banzaiUtils.LogErrorf(banzaiConstants.TagFetchClusterConfig, "Error during get service client %v", err)
+		return nil, getBanzaiErrorFromError(err)
+	}
+	banzaiUtils.LogInfo(banzaiConstants.TagFetchClusterConfig, "Get Google Service Client succeeded")
+
+	banzaiUtils.LogInfof(banzaiConstants.TagFetchClusterConfig, "Get google cluster with name %s", cs.Name)
+	cl, err := svc.Projects.Zones.Clusters.Get(cs.Google.Project, cs.Location, cs.Name).Context(context.Background()).Do()
```
Comment (baluchicken) [inline cloud/gke.go]: This can be replaced with GetClusterGoogle()

File: conf/config.toml.example [hunk 30]
```diff
@@ -30,6 +30,9 @@ clientsecret = ""
 # AccessToken settings
 # tokensigningkey = ""
 
+# GKE credential path
```
Comment (baluchicken) [inline conf/config.toml.example]: This parameter is here multiple times remove one of them

File: main.go [hunk 31]
```diff
@@ -410,6 +412,26 @@ func CreateCluster(c *gin.Context) {
 				cloud.JsonKeyMessage: err,
 			})
 		}
+	case banzaiConstants.Google:
+		// validate and create Azure cluster
```
Comment (baluchicken) [inline main.go]: wrong comment this one is Google cluster

File: cloud/gke.go (hunk 28 above)
Comment (colin014) [reply cloud/gke.go]: It's the Delete function, not Get

Comment (baluchicken) [reply cloud/gke.go]: Ohh okey, you are right sorry my mistake:)

**Comment by baluchicken** [comment]
LGTM thx @colin014 :)

**Pull request closed without merging**